src/Http/Http/src/StreamPipeReader.cs

```diff
@@ -0,0 +1,368 @@
+                {
+                    AllocateReadTail();
+#if NETCOREAPP2_2
+                    var length = await _readingStream.ReadAsync(_readTail.AvailableMemory, tokenSource.Token);
+#elif NETSTANDARD2_0
+                    if (!MemoryMarshal.TryGetArray<byte>(_readTail.AvailableMemory, out var arraySegment))
+                    {
+                        ThrowHelper.ThrowInvalidCastException_NoArrayFromMemory();
+                    }
+
+                    var length = await _readingStream.ReadAsync(arraySegment.Array, 0, arraySegment.Count, tokenSource.Token);
+#else
+#error Target frameworks need to be updated.
+#endif
+                    _readTail.End += length;
+                    _consumedLength += length;
+                }
+                catch (OperationCanceledException)
+                {
+                    ClearOutCancellation();
+
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        throw;
+                    }
+
+                    isCanceled = true;
+                }
+
+                return new ReadResult(GetCurrentReadOnlySequence(), isCanceled, IsCompletedOrThrow());
+            }
+        }
+
+        private void ClearOutCancellation() => Interlocked.Exchange(ref _internalTokenSource, null);
+
+        private void ThrowIfCompleted()
+        {
+            if (_isCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoReadingAllowed();
+            }
+        }
+
+        public override bool TryRead(out ReadResult result)
+        {
+            ThrowIfCompleted();
+
+            return TryReadInternal(InternalTokenSource, out result);
+        }
+
+        private bool TryReadInternal(CancellationTokenSource source, out ReadResult result)
+        {
+            var isCancellationRequested = source.IsCancellationRequested;
+            if (isCancellationRequested || _consumedLength > 0 && !_examinedEverything)
+            {
+                // If TryRead/ReadAsync are called and cancellation is requested, we need to make sure memory is allocated for the ReadResult,
+                // otherwise if someone calls advance afterward on the ReadResult, it will throw.
+                if (isCancellationRequested)
+                {
+                    AllocateReadTail();
+
+                    ClearOutCancellation();
+                }
+
+                result = new ReadResult(
+                    GetCurrentReadOnlySequence(),
+                    isCanceled: isCancellationRequested,
+                    IsCompletedOrThrow());
+                return true;
+            }
+
+            result = new ReadResult();
+            return false;
+        }
+
+        private ReadOnlySequence<byte> GetCurrentReadOnlySequence()
+        {
+            return new ReadOnlySequence<byte>(_readHead, _readIndex, _readTail, _readTail.End - _readTail.Start);
+        }
+
+        private void AllocateReadTail()
```

Comment:
File an issue to have a MaxBufferSize setting

Reply:
Will file once merged.